doc/userguide/firewall/firewall-design.rst

```diff
@@ -0,0 +1,158 @@
+
+* ``request_not_started``
+* ``request_line``
+* ``request_headers``
+* ``request_body``
+* ``request_trailer``
+* ``request_complete``
+
+Response (``to_client``) side:
+
+* ``response_not_started``
+* ``response_line``
+* ``response_headers``
+* ``response_body``
+* ``response_trailer``
+* ``response_complete``
+
+tls
+^^^
+
+Available states:
+
+Request (``to_server``) side:
+
+* ``client_in_progress``
+* ``client_hello_done``
+* ``client_cert_done``
+* ``client_handshake_done``
+* ``client_finished``
+
+Response (``to_client``) side:
+
+* ``server_in_progress``
+* ``server_hello``
+* ``server_cert_done``
+* ``server_hello_done``
+* ``server_handshake_done``
+* ``server_finished``
+
+Firewall pipeline
+~~~~~~~~~~~~~~~~~
+
+The firewall pipeline works in the detection engine, and is invoked after packet decoding, flow
+update, stream tracking and reassembly and app-layer parsing are all done in the context of a
+single packet.
+
+For each packet rules in the first firewall hook ``packet_filter`` are then evaluated. Assuming
+the verdict of this hook is ``accept:hook``, the next hook is evaluated: ``packet_td`` (packet
+threat detection). In this hook the IDS/IPS rules are evaluated. Rule actions here are not
+immediate, as they can still be modified by alert postprocessing like rate_filter, thresholding, etc.
+
+If the packet has been marked internally as a packet with an application layer update, then the
+next table is ``app_filter``.
+
+In ``app_filter`` the per application layer states are all evaluated at least once. At each of
+these states an ``accept:hook`` required to progress to the next state. When all available states
+have been accepted, the pipeline moves to the final table ``app_td`` (application layer threat
+detection). A ``drop`` in the ``app_filter`` table is immediate, however and ``accept`` is
+conditional on the verdict of the ``app_td`` table.
+
+In ``app_td`` the IDS/IPS rules for the application layer are evaluated. ``drop`` actions in this
+table are queued in the alert queue.
+
+Then all tables have been evaluated, the alert finalize process orders threat detection alerts
+by ``action-order`` logic. It can then apply a ``drop`` or default to ``accept``-ing.
+
+
+.. image:: fw-pipeline.png
+
+
+Pass rules with Firewall mode
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In IDS/IPS mode, a ``pass`` rule with app-layer matches will bypass the detection engine for the
+rest of the flow. In firewall mode, this bypass no longer happens in the same way, as ``pass`` rules
+do not affect firewall rules. So the detection engine is still invoked on packets of such a flow,
+but the ``packet_td`` and ``app_td`` tables are skipped.
+
+
+Concepts
+~~~~~~~~
+
+* ``table`` collection of rules with different properties: ``packet_filter``, ``packet_td``,
+  ``app_filter``, ``app_td``.
+* ``hook``  / ``state`` controls a specific state at which a rule is evaluated. Examples are
+  ``tcp.flow_start`` or ``tls.client_body_done``.
```

Comment:
Considering all that is being covered here is new, maybe this could come before the explanation?

Reply:
Yeah makes sense